src/ray/raylet/scheduling/local_task_manager.cc

```diff
@@ -0,0 +1,1061 @@
+        bool did_spill = TrySpillback(work, is_infeasible);
+        if (!did_spill) {
+          // There must not be any other available nodes in the cluster, so the task
+          // should stay on this node. We can skip the rest of the shape because the
+          // scheduler will make the same decision.
+          work->SetStateWaiting(
+              internal::UnscheduledWorkCause::WAITING_FOR_RESOURCES_AVAILABLE);
+          break;
+        }
+        if (!spec.GetDependencies().empty()) {
+          task_dependency_manager_.RemoveTaskDependencies(
+              task.GetTaskSpecification().TaskId());
+        }
+        work_it = dispatch_queue.erase(work_it);
+      } else {
+        // Force us to recalculate the next update time the next time a task
+        // comes through this queue. We should only do this when we're
+        // confident we're ready to dispatch the task after all checks have
+        // passed.
+        sched_cls_info.next_update_time = std::numeric_limits<int64_t>::max();
+        sched_cls_info.running_tasks.insert(spec.TaskId());
+        // The local node has the available resources to run the task, so we should run
+        // it.
+        std::string allocated_instances_serialized_json = "{}";
+        if (RayConfig::instance().worker_resource_limits_enabled()) {
+          allocated_instances_serialized_json =
+              cluster_resource_scheduler_->GetLocalResourceManager()
+                  .SerializedTaskResourceInstances(allocated_instances);
+        }
+        work->allocated_instances = allocated_instances;
+        work->SetStateWaitingForWorker();
+        bool is_detached_actor = spec.IsDetachedActor();
+        auto &owner_address = spec.CallerAddress();
+        /// TODO(scv119): if a worker is not started, the resources is leaked and
+        // task might be hanging.
+        worker_pool_.PopWorker(
+            spec,
+            [this, task_id, scheduling_class, work, is_detached_actor, owner_address](
+                const std::shared_ptr<WorkerInterface> worker, PopWorkerStatus status,
+                const std::string &runtime_env_setup_error_message) -> bool {
+              return PoppedWorkerHandler(worker, status, task_id, scheduling_class, work,
+                                         is_detached_actor, owner_address,
+                                         runtime_env_setup_error_message);
+            },
+            allocated_instances_serialized_json);
+        work_it++;
+      }
+    }
+    // In the beginning of the loop, we add scheduling_class
+    // to the `info_by_sched_cls_` map.
+    // In cases like dead owners, we may not add any tasks
+    // to `running_tasks` so we can remove the map entry
+    // for that scheduling_class to prevent memory leaks.
+    if (sched_cls_info.running_tasks.size() == 0) {
+      info_by_sched_cls_.erase(scheduling_class);
+    }
+    if (is_infeasible) {
+      // TODO(scv119): fail the request.
```

Comment:
synced with Chen. It seems that this should never happen. Maybe let's just RAY_CHECK here in case of any failure.

Reply:
Just curious, but why is it a bug? Is this code path not reachable when pg is removed?